modules/core/src/com/haulmont/cuba/core/app/EntitySnapshotManager.java

```diff
@@ -336,4 +315,95 @@ protected String processSnapshotXml(String snapshotXml, Map<Class, Class> classM
         replaceInXmlTree(document.getRootElement(), classMapping);
         return document.asXML();
     }
+
+    @Override
+    public EntitySnapshot getLastEntitySnapshot(Entity entity) {
+        MetaClass metaClass = entity.getMetaClass();
+        View view = metadata.getViewRepository().getView(EntitySnapshot.class, "entitySnapshot.browse");
+        LoadContext<EntitySnapshot> lx = LoadContext.create(EntitySnapshot.class).setQuery(LoadContext
+                .createQuery(format("select e from sys$EntitySnapshot e where e.entityMetaClass = :metaClass and"
+                                + " e.entity.%s = :entityId order by e.createTs desc",
+                        referenceToEntitySupport.getReferenceIdPropertyName(metaClass)))
+                .setParameter("metaClass", metaClass.getName())
+                .setParameter("entityId", entity.getId())
+                .setMaxResults(1))
+                .setView(view);
+        return dataManager.load(lx);
+    }
+
+    @Override
+    public EntitySnapshot getLastEntitySnapshot(MetaClass metaClass, Object id) {
+        metaClass = extendedEntities.getOriginalOrThisMetaClass(metaClass);
+        Entity entity = dataManager.load(new LoadContext<>(metaClass).setId(id).setView(View.LOCAL));
```

Comment:
Let's use single transaction in this method. Do not mix dataManager and Persistence.